src/circuit/gadget/poseidon.rs

```diff
@@ -91,170 +101,176 @@ impl<
     }
 }
 
-fn poseidon_duplex<
+fn poseidon_sponge<
     F: FieldExt,
-    PoseidonChip: PoseidonDuplexInstructions<F, S, T, RATE>,
+    PoseidonChip: PoseidonSpongeInstructions<F, S, D, T, RATE>,
     S: Spec<F, T, RATE>,
-    D: Domain<F, T, RATE>,
+    D: Domain<F, RATE>,
     const T: usize,
     const RATE: usize,
 >(
     chip: &PoseidonChip,
     mut layouter: impl Layouter<F>,
-    domain: &D,
     state: &mut State<PoseidonChip::Word, T>,
-    input: &SpongeState<PoseidonChip::Word, RATE>,
-) -> Result<SpongeState<PoseidonChip::Word, RATE>, Error> {
-    *state = chip.pad_and_add(&mut layouter, domain, state, input)?;
+    input: Option<&Absorbing<PaddedWord<F>, RATE>>,
+) -> Result<Squeezing<PoseidonChip::Word, RATE>, Error> {
+    if let Some(input) = input {
+        *state = chip.add_input(&mut layouter, state, input)?;
+    }
     *state = chip.permute(&mut layouter, state)?;
     Ok(PoseidonChip::get_output(state))
 }
 
-/// A Poseidon duplex sponge.
+/// A Poseidon sponge.
 #[derive(Debug)]
-pub struct Duplex<
+pub struct Sponge<
     F: FieldExt,
-    PoseidonChip: PoseidonDuplexInstructions<F, S, T, RATE>,
+    PoseidonChip: PoseidonSpongeInstructions<F, S, D, T, RATE>,
     S: Spec<F, T, RATE>,
-    D: Domain<F, T, RATE>,
+    M: SpongeMode,
+    D: Domain<F, RATE>,
     const T: usize,
     const RATE: usize,
 > {
     chip: PoseidonChip,
-    sponge: Sponge<PoseidonChip::Word, RATE>,
+    mode: M,
     state: State<PoseidonChip::Word, T>,
-    domain: D,
+    _marker: PhantomData<(M, D)>,
```

Comment:
Do we need `M` in this `PhantomData`?

```suggestion
    _marker: PhantomData<D>,
```

Reply:
I did need it in an earlier version, but at some point I refactored the sponge mode to store the mode state instead of being purely a type marker. So yes, we don't need it here.